File: 201-250/226_invert_binary_tree_iterative_dfs.go
// Exercise: https://leetcode.com/problems/invert-binary-tree/description/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func invertTree(root *TreeNode) *TreeNode {
    if root == nil {
        return nil;
    }
    stack := make([]*TreeNode, 0, 16)
    stack = append(stack, root)
    for len(stack) > 0 {
        lastIdx := len(stack)-1
        node := stack[lastIdx]
        stack = stack[:lastIdx]
        if node == nil {
            continue
        }
        node.Left, node.Right = node.Right, node.Left
        if node.Left != nil {
            stack = append(stack, node.Left)
        }
        if node.Right != nil {
            stack = append(stack, node.Right)
        }
    }
    return root
}

File: 201-250/226_invert_binary_tree_iterative_dfs.java
import java.util.Deque;
import java.util.ArrayDeque;

// Problem: https://leetcode.com/problems/invert-binary-tree/description/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;
        Deque<TreeNode> stack = new ArrayDeque<>();
        stack.push(root);
        while (stack.size() > 0) {
            TreeNode node = stack.pop();
            if (node == null) continue;
            TreeNode t = node.right;
            node.right = node.left;
            node.left = t;
            if (node.left != null) {
                stack.push(node.left);
            }
            if (node.right != null) {
                stack.push(node.right);
            }
        }
        return root;
    }
}
